src/server.rs

```diff
@@ -383,29 +383,32 @@ async fn index(
 
     // Before middleware
     // Global
-    request = middleware_router
-        .get_global_middlewares(&MiddlewareType::BeforeRequest)
-        .iter()
-        .fold(request, |request, function| {
-            // TODO: parallelize in the future
-            block_on(execute_middleware_function(&request, function)).unwrap_or_else(|e| {
-                error!("Error while executing before middleware: {e}");
-                request
-            })
-        });
+    let mut before_middlewares =
+        middleware_router.get_global_middlewares(&MiddlewareType::BeforeRequest);
     // Route specific
-    request = match middleware_router.get_route(&MiddlewareType::BeforeRequest, req.uri().path()) {
-        Some((function, route_params)) => {
-            request.path_params = route_params;
-            execute_middleware_function(&request, &function)
-                .await
-                .unwrap_or_else(|e| {
-                    error!("Error while executing before middleware: {e}");
-                    request
-                })
-        }
-        None => request,
-    };
+    if let Some((function, route_params)) =
+        middleware_router.get_route(&MiddlewareType::BeforeRequest, req.uri().path())
+    {
+        before_middlewares.push(function);
+        request.path_params = route_params;
+    }
+    for before_middleware in before_middlewares {
+        request = match execute_middleware_function(&request, &before_middleware).await {
+            Ok(MiddlewareReturn::Request(r)) => r,
+            Ok(MiddlewareReturn::Response(r)) => {
+                // If a before middleware returns a response, we abort the request and return the response
```

Comment:
I think we should ideally be propagating the response to the next middleware, no?

Reply:
The input of a before middleware is a `Request` object, not a `Response`. I don't think we can send it to the next middleware. I have some ideas to implement middleware chaining in the future but I think if one of the before middleware returns a `Response` on the way, we should abort the request ?
We could only pass the `Response` to the after middlewares because that is what they use as an input, but I am not sure we should do this as we may want to consider the user has be authenticated (which would be done in a before middleware) when the request reaches the after middleware if that makes sense.

Reply:
I was thinking of implementing something like this (https://emmett.sh/docs/2.2.x/pipeline#pipeline ) when I was talking about middleware chaining. We can have a `send_to_next` and `recieved_from_prev` arguments or similar?

Reply:
Thank you for sharing this it is very interesting! I think what he calls pipe is the equivalent of our before middleware. We could have something similar but that would require reworking the whole middleware architecture. I think it is outside the scope of this PR but this article can be very valuable to implement a miiddleware chaining solution later :)

Reply:
>  I think it is outside the scope of this PR but this article can be very valuable to implement a miiddleware chaining solution later :)

Yes, absolutely. I agree 😄